flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1618,17 +1729,166 @@ def test_run(self, logger):
+    )
+):
+    """
+    Tests for ``Mountblockdevice`` initialization.
+    """
+
+
+class MountBlockDeviceTests(
+    make_istatechange_tests(
+        MountBlockDevice,
+        dict(volume=_ARBITRARY_VOLUME, mountpoint=FilePath(b"/foo")),
+        dict(volume=_ARBITRARY_VOLUME, mountpoint=FilePath(b"/bar")),
+    )
+):
     """
-    return UnmountBlockDevice(
-        volume=_ARBITRARY_VOLUME,
+    Tests for ``MountBlockDevice``\ 's ``IStateChange`` implementation.
+    """
+    def _run_test(self, mountpoint):
+        """
+        Verify that ``MountBlockDevice.run`` mounts the filesystem from the
+        block device for the attached volume it is given.
+        """
+        host = u"192.0.7.8"
+        dataset_id = uuid4()
+        api = loopbackblockdeviceapi_for_test(self)
+        volume = api.create_volume(
+            dataset_id=dataset_id, size=REALISTIC_BLOCKDEVICE_SIZE,
+        )
+        api.attach_volume(volume.blockdevice_id, host)
+
+        deployer = BlockDeviceDeployer(
+            hostname=host,
+            block_device_api=api,
+            mountroot=mountpoint.parent(),
+        )
+
+        filesystem = u"ext4"
+        self.successResultOf(
+            CreateFilesystem(volume=volume, filesystem=filesystem).run(
+                deployer
+            )
+        )
+
+        change = MountBlockDevice(volume=volume, mountpoint=mountpoint)
+        self.successResultOf(run_state_change(change, deployer))
+
+        expected = (
+            api.get_device_path(volume.blockdevice_id).path,
+            mountpoint.path,
+            filesystem,
+        )
+        mounted = list(
+            (part.device, part.mountpoint, part.fstype)
+            for part in psutil.disk_partitions()
+        )
+        self.assertIn(expected, mounted)
+
+    def test_run(self):
+        """
+        ``CreateFilesystem.run`` initializes a block device with a filesystem
+        which ``MountBlockDevice.run`` can then mount.
+        """
+        mountroot = mountroot_for_test(self)
+        mountpoint = mountroot.child(b"mount-test")
+        self._run_test(mountpoint)
+
+    def test_mountpoint_exists(self):
+        """
+        It is not an error if the mountpoint given to ``MountBlockDevice``
+        already exists.
+        """
+        mountroot = mountroot_for_test(self)
+        mountpoint = mountroot.child(b"mount-test")
+        mountpoint.makedirs()
+        self._run_test(mountpoint)
+
+    def test_mountpoint_error(self):
+        """
+        If the mountpoint is unusable, for example because it is a regular file
+        instead of a directory, ``MountBlockDevice.run`` returns a ``Deferred``
+        that fires with a ``Failure`` given the reason.
+        """
+        mountroot = mountroot_for_test(self)
+        intermediate = mountroot.child(b"mount-error-test")
+        intermediate.setContent(b"collision")
+        mountpoint = intermediate.child(b"mount-test")
+        exception = self.assertRaises(
```

Comment:
Or rather, you aren't proving a `Deferred` was returned. Maybe `successResultof` unwrapped it (I forget how it works) or maybe it got raised.